src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs

```diff
@@ -1156,6 +1175,106 @@ private BoundExpression CreateAndPopulateList(BoundCollectionExpression node, Ty
                 collectionType);
         }
 
+        /// <summary>
+        /// Create and populate a dictionary from a collection expression.
+        /// The collection may or may not have a known length.
+        /// </summary>
+        private BoundExpression CreateAndPopulateDictionary(BoundCollectionExpression node, ImmutableArray<TypeWithAnnotations> typeArguments, ImmutableArray<BoundNode> elements)
+        {
+            Debug.Assert(!_inExpressionLambda);
+
+            var collectionType = _factory.WellKnownType(WellKnownType.System_Collections_Generic_Dictionary_KV).Construct(typeArguments);
+
+            var localsBuilder = ArrayBuilder<BoundLocal>.GetInstance();
```

Comment:
I'll leave this as `BoundLocal` since it matches other methods in this file that track, and it's likely we'll need `BoundLocal` instances when we start relying on helper methods such as `RewriteCollectionExpressionElementsIntoTemporaries`.